BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift

```diff
@@ -0,0 +1,192 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveRewards
+
+protocol NotificationHandlerDelegate: AnyObject {
+  func show(notificationView: WalletNotificationView)
+  func handleAction(notification: RewardsNotification)
+  func hideNotifications()
+}
+
+class NotificationHandler {
```

Comment:
Overall, I dont understand this class well…

Its acting as a receiver to turn `RewardsNotification` into one of the `Wallet*Notfication` structs (View Model), in which it then tells the delegate to show notifications (Controller) by creating a view for it (View Controller) and at the same time handle its own notification stack (Model) by clearing away things automatically.

IMO, this class shouldn't be doing all of this, and we should be picking a specific direction. Do we want it to be a "view model store" of sorts, turning `RewardsNotification`s into something that WalletViewController can read and determine what to show?

Whats stopping us of just making this class hold a list of enums that look like this:

```swift
enum WalletNotification
  case alert(AlertNotification)
  case action(ActionNotification)
  case message(MessageNotification)
}
```

On init it just looks at all ledger notifications, converts them, and adds them to its list. At the same time it subscribes to BATBraveLedgerNotificationAdded. Then you can just add simple ledger wrapper methods here that control the list (add, clear, etc.) 

And `WalletViewController` just reads that list in `viewDidLoad` and shows the first one? Then when the user clicks on something we can call `notificationStore.clear(walletNotification)` which will then call the appropriate `BATBraveLedger` method

The only thing that the notification store would still have to do is bind the `WalletViewController` to when the list updates through a delegate method.

I also don't think we should be not subscribing to changes between the point of `init` and `viewDidLoad`. That introduces a small race condition between those two points where we initially read out the notifications in the ledger and not receiving a new notification. This would be fixed by subscribing immediately on init and not forcing the `WalletViewController` to show some view when the list changes even if the view hasn't loaded yet. It'd do it itself in viewDidLoad by just reading out the notification models.

Let me know if we need further clarification around this or if I'm missing something that breaks this thought

Reply:
Yeah so its basically a helper class that generates views for notifications and tells that to controller. No stack is maintained in this class, as doing so would require implementing other observers like `delete` etc from braveLedger. 

I created this class to keep WalletViewController decoupled from notification logic and also keep the number of lines in check. But I think the helper class can be further separated into more solid structures. 

I have pushed some refactors. With these changes it gets more clearer.
We have three options from it:
1. Keep as is
2. Let RewardsNotificationService forward notification and not notificationView to WalletViewController and let WalletViewController manage rest.
3. Shift all the code in RewardsNotificationService into WalletViewController.

My reason was that WalletViewController does not depend on Notifications at all. In fact the `notificationView` & `setNotificationView` in `WalletView` are also redundant(Can be simple 
 UIView). So making this class gives us the option to add/remove this feature like an addon.

I am still looking for a better way to do this, let me know what you think is better.

Reply:
> I created this class to keep WalletViewController decoupled from notification logic

This is why I'm partially confused though because most of the logic is already decoupled because it's in `BATBraveLedger`. It has a stack, of which is managed, and you can call clear on, so all you're doing is abstracting flow logic between `RewardsNotification` and a few different view's per type of `RewardsNotification` from `WalletViewController` into a class which is only used by `WalletViewController` and is unlikely to be used anywhere else since its view specific.

And if the conversion of models to view models is what you want to pull away from WalletViewController thats fine, but the rest of the flow doesn't need to be in there (load next, close button, etc). It could just be in `WalletViewController`

If line count is a big deal you could create a new swift file that just extends `WalletViewController` and call it `WalletViewController+Notifications` or something

> My reason was that WalletViewController does not depend on Notifications at all. In fact the notificationView & setNotificationView in WalletView are also redundant(Can be simple UIView)

This doesn't have to be set in stone though, since again, these notifications are shown nowhere else.  You could just make a `UIViewController` subclass that replaces the NotificationView stuff in WalletViewController.  Then you are essentially only adding logic in `WalletViewController` on whether or not to show said controller based on non-empty notifications in the ledger and maybe getting a _did add notification_ notification which would prompt a check to notifications.

Reply:
Cool, I moved the code into WalletViewController.